flare/algorithm_zoo/ppo.py

```diff
@@ -0,0 +1,81 @@
+from flare.framework.algorithm import Algorithm
+from flare.framework import common_functions as comf
+from torch.distributions import Categorical
+import torch
+import numpy as np
+from copy import deepcopy
+
+
+class PPO2(Algorithm):
+    """
+    PPO2 of the clipped surrogate objective
+
+    learn() requires keywords: "action", "reward", "v_value", "action_log_prob"
+    """
+
+    def __init__(self,
+                 model,
+                 iterations_per_batch,
+                 epsilon=0.1,
+                 gpu_id=-1,
+                 discount_factor=0.99,
+                 value_cost_weight=1.0):
+
+        super(PPO2, self).__init__(model, gpu_id, iterations_per_batch)
+        self.epsilon = epsilon
+        self.discount_factor = discount_factor
+        self.value_cost_weight = value_cost_weight
+
+    def get_action_specs(self):
+        ### "action_log_prob" is required by the algorithm but not by the model
+        return self.model.get_action_specs() + [
+            ("action_log_prob", dict(shape=[1]))
+        ]
+
+    def learn(self, inputs, next_inputs, states, next_states, next_episode_end,
+              actions, next_actions, rewards):
+        """
+        This learn() is expected to be called multiple times on each minibatch
+        """
+
+        action = actions["action"]
+        log_prob = actions["action_log_prob"]
+        reward = rewards["reward"]
+
+        values, states_update = self.model.value(inputs, states)
+        value = values["v_value"]
+
+        with torch.no_grad():
+            next_values, next_states_update = self.model.value(next_inputs,
+                                                               next_states)
+            next_value = next_values["v_value"] * (
+                1 - next_episode_end["next_episode_end"])
+
+        assert value.size() == next_value.size()
+
+        critic_value = reward + self.discount_factor * next_value
+        td_error = (critic_value - value).squeeze(-1)
```

Comment:
PPO used a set of pre-computed td errors. Maybe we need a hook function that is called before `learn()`?

Reply:
This is what I said in my comments. The original version does use pre-computed advantages. For convenience, I compute them on the fly with the current parameters. I think the original way is an approximation (not correct in theory). Even if we want to do that, we can always output advantages as part of the actions, and then use them as inputs for learn().